File: platform/jvm/compiler/flabbergast/MainPrinter.java
package flabbergast;

import java.io.File;
import java.io.IOException;
import java.util.EnumSet;
import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.CommandLineParser;
import org.apache.commons.cli.GnuParser;
import org.apache.commons.cli.HelpFormatter;
import org.apache.commons.cli.Options;
import org.apache.commons.cli.ParseException;

public class MainPrinter {
  public static void main(String[] args) {
    Options options = new Options();
    options.addOption("o", "output", true, "Write output to file instead of standard output.");
    options.addOption("t", "trace-parsing", false, "Produce a trace of the parse process.");
    options.addOption("p", "no-precomp", false, "Do not use precompiled libraries");
    options.addOption("h", "help", false, "Show this message and exit");
    CommandLineParser cl_parser = new GnuParser();
    CommandLine result;

    try {
      result = cl_parser.parse(options, args);
    } catch (ParseException e) {
      System.err.println(e.getMessage());
      System.exit(1);
      return;
    }

    if (result.hasOption('h')) {
      HelpFormatter formatter = new HelpFormatter();
      System.err.println("Run a Flabbergast file and display the “value” attribute.");
      formatter.printHelp("gnu", options);
      System.exit(1);
    }

    String[] files = result.getArgs();
    if (files.length != 1) {
      System.err.println("Exactly one Flabbergast script must be given.");
      System.exit(1);
    }

    ResourcePathFinder resource_finder = new ResourcePathFinder();
    try {
      resource_finder.prependPath(
          new File(new File(files[0]).getParentFile(), "lib").getCanonicalPath());
    } catch (IOException e) {
      e.printStackTrace();
    }
    resource_finder.addDefaults();
    ErrorCollector collector = new ConsoleCollector();
    ConsoleTaskMaster task_master = new ConsoleTaskMaster();
    DynamicCompiler compiler = new DynamicCompiler(collector);
    compiler.setFinder(resource_finder);
    task_master.addUriHandler(compiler);
    EnumSet<LoadRule> rules = EnumSet.noneOf(LoadRule.class);
    if (result.hasOption('p')) {
      rules.add(LoadRule.PRECOMPILED);
    }
    task_master.addAllUriHandlers(resource_finder, rules);
    try {
      Parser parser = Parser.open(files[0]);
      parser.setTrace(result.hasOption('t'));
      Class<? extends Future> run_type =
          parser.parseFile(collector, compiler.getCompilationUnit(), "Printer");
      if (run_type != null) {
        Future computation = run_type.getConstructor(TaskMaster.class).newInstance(task_master);
        PrintResult filewriter =
            new PrintResult(task_master, computation, result.getOptionValue('o'));
        filewriter.slot();
        task_master.run();
        task_master.reportCircularEvaluation();
        System.exit(filewriter.getSuccess() ? 0 : 1);
      }
    } catch (Exception e) {
      System.err.println(e.getMessage());
      e.printStackTrace();
    }
    System.exit(1);
  }
}

File: platform/cli/main-printer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Reflection;
using System.Reflection.Emit;
using Flabbergast;
using NDesk.Options;

public class Printer {
    public static int Main(string[] args) {
        var trace = false;
        string output_filename = null;
        var show_help = false;
        var use_precompiled = true;
        var options = new OptionSet {
            {"o=|output", "Write output to file instead of standard output.", v => output_filename = v},
            {"t|trace-parsing", "Produce a trace of the parse process.", v => trace = v != null},
            {"p|no-precomp", "do not use precompiled libraries", v => use_precompiled = v == null},
            {"h|help", "show this message and exit", v => show_help = v != null}
        };

        List<string> files;
        try {
            files = options.Parse(args);
        } catch (OptionException e) {
            Console.Error.Write(AppDomain.CurrentDomain.FriendlyName + ": ");
            Console.Error.WriteLine(e.Message);
            Console.Error.WriteLine("Try “" + AppDomain.CurrentDomain.FriendlyName + " --help” for more information.");
            return 1;
        }

        if (show_help) {
            Console.WriteLine("Usage: " + AppDomain.CurrentDomain.FriendlyName + " input.o_0");
            Console.WriteLine("Run a Flabbergast file and display the “value” attribute.");
            Console.WriteLine();
            Console.WriteLine("Options:");
            options.WriteOptionDescriptions(Console.Out);
            return 1;
        }

        if (files.Count != 1) {
            Console.Error.WriteLine("Exactly one Flabbergast script must be given.");
            return 1;
        }

        var assembly_builder = AppDomain.CurrentDomain.DefineDynamicAssembly(new AssemblyName("Print"), AssemblyBuilderAccess.Run);
        CompilationUnit.MakeDebuggable(assembly_builder);
        var module_builder = assembly_builder.DefineDynamicModule("PrintModule", true);
        var unit = new CompilationUnit(module_builder, true);
        var collector = new ConsoleCollector();
        var task_master = new ConsoleTaskMaster();
        var resource_finder = new ResourcePathFinder();
        resource_finder.PrependPath(Path.Combine(Path.GetDirectoryName(Path.GetFullPath(files[0])), "lib"));
        resource_finder.AddDefault();
        task_master.AddUriHandler(new CurrentInformation(false));
        task_master.AddUriHandler(BuiltInLibraries.INSTANCE);
        var db_handler = new DbUriHandler();
        db_handler.Finder = resource_finder;
        task_master.AddUriHandler(db_handler);
        task_master.AddUriHandler(EnvironmentUriHandler.INSTANCE);
        task_master.AddUriHandler(HttpHandler.INSTANCE);
        task_master.AddUriHandler(FtpHandler.INSTANCE);
        task_master.AddUriHandler(FileHandler.INSTANCE);
        var resource_handler = new ResourceHandler();
        resource_handler.Finder = resource_finder;
        task_master.AddUriHandler(resource_handler);
        if (use_precompiled) {
            var precomp = new LoadPrecompiledLibraries();
            precomp.Finder = resource_finder;
            task_master.AddUriHandler(precomp);
        }
        var dyncomp = new DynamicallyCompiledLibraries(collector);
        dyncomp.Finder = resource_finder;
        task_master.AddUriHandler(dyncomp);
        var parser = Parser.Open(files[0]);
        parser.Trace = trace;
        var run_type = parser.ParseFile(collector, unit, "Printer");
        if (run_type != null) {
            var computation = (Computation) Activator.CreateInstance(run_type, task_master);
            var filewriter = new PrintResult(task_master, computation, output_filename);
            filewriter.Slot();
            task_master.Run();
            task_master.ReportCircularEvaluation();
            return filewriter.Success ? 0 : 1;
        }
        return 1;
    }
}
